tfhe/src/integer/server_key/radix_parallel/count_zeros_ones.rs

```diff
@@ -0,0 +1,543 @@
+        // must be chunk_exact by 3 otherwise there was an error earlier
+        assert_eq!(duo_completer_blocks.len() % 3, 0);
+
+        let lut_count_bits_full_range = self.key.generate_lookup_table(|x| {
+            let mut count = 0u64;
+            for i in 0..(num_bits_in_block * 2) {
+                count += count_kind.is_ok((x >> i) & 1);
+            }
+            count
+        });
+
+        let lut_count_bits_half_range = self.key.generate_lookup_table(|x| {
+            let mut count = 0u64;
+            for i in 0..num_bits_in_block {
+                count += count_kind.is_ok((x >> i) & 1);
+            }
+            count
+        });
+
+        let (bit_count_of_packed_blocks, mut bit_counts_of_completer_blocks) = rayon::join(
+            || {
+                regular_blocks
+                    .par_chunks_exact(2)
+                    .map(|chunk_of_two| {
+                        let mut packed = self.pack_block_chunk(chunk_of_two);
+                        self.key
+                            .apply_lookup_table_assign(&mut packed, &lut_count_bits_full_range);
+                        packed
+                    })
+                    .collect::<Vec<_>>()
+            },
+            || {
+                let luts = [
+                    self.key.generate_lookup_table(|x| {
+                        let mut count = 0u64;
+                        for i in 0..num_bits_in_block + 1 {
+                            count += count_kind.is_ok((x >> i) & 1);
+                        }
+                        count
+                    }),
+                    self.key.generate_lookup_table(|x| {
+                        let mut count = 0u64;
+                        for i in 1..num_bits_in_block + 2 {
+                            count += count_kind.is_ok((x >> i) & 1);
+                        }
+                        count
+                    }),
+                ];
+
+                let mut packed_blocks = Vec::new();
+                for chunk_of_3 in duo_completer_blocks.chunks_exact(3) {
+                    packed_blocks.push(self.pack_block_chunk(&chunk_of_3[..2]));
+                    packed_blocks.push(self.pack_block_chunk(&chunk_of_3[1..3]));
+                }
+
+                packed_blocks
+                    .par_iter()
+                    .enumerate()
+                    .map(|(i, packed_block)| {
+                        self.key.apply_lookup_table(packed_block, &luts[(i) % 2])
+                    })
+                    .chain(single_completer_blocks.par_iter().map(|block| {
+                        self.key
+                            .apply_lookup_table(block, &lut_count_bits_half_range)
+                    }))
+                    .collect::<Vec<_>>()
+            },
+        );
+
+        let num_sum = ((self.message_modulus().0 * self.carry_modulus().0) - 1)
+            / (self.message_modulus().0 * self.carry_modulus().0).ilog2() as usize;
```

Comment:
not sure I follow the ilog2 at the denominator location ?

Reply:
I'ts because each block we want to sum, contains the count_zeros/ones of 2 blocks so the value is 0..total_num_bits <=> 0..(self.message_modulus().0 * self.carry_modulus().0).ilog2()

so 15 / 4 in 2_2 = > 3, to clearer  and 'safer' (not really since 3 < 5), maybe we can wait for the safe sum to be merged and rewrite as:

```rust
let bit_count_of_packed_block_degree =  4; // Since each block encrypts 4 bits, max count is 4
let num_sum = self.max_sum_size(bit_count_of_packed_block_degree);
```

Reply:
I'm still a bit unclear, to me we have an integer value being divided by a number of bits, hence something odd in terms of units

Reply:
a number of bits is also an integer value, here we want to know how many value of Degree(4) we can sum at once

Reply:
yep ok said this way I have a better time understanding.

as for the safe sum up to you